Since solicited node and IPv4 link local multicast MDB group is not created, we
exclude optimized router forwarding for solicited node and link local multicast
packets

diff --git a/net/bridge/br_multicast.c b/net/bridge/br_multicast.c
index b16f5cd..000cc59 100644
--- a/net/bridge/br_multicast.c
+++ b/net/bridge/br_multicast.c
@@ -1682,6 +1682,23 @@ err_out:
 }
 
 #if IS_ENABLED(CONFIG_IPV6)
+/* Check if addr is  solicited node multicast FF02::1:FFxx:xxxx
+ */
+static inline int is_solicited_node_mcast(const struct in6_addr *addr)
+{
+	return ((((addr->s6_addr32[0] ^ htonl(0xff020000)) |
+		addr->s6_addr32[1] | (addr->s6_addr32[2] ^ htonl(0x00000001))) == 0)
+		&& (addr->s6_addr[12] == 0xff));
+}
+
+/* Check if addr is link local multicast FF02::xx
+ */
+static inline int is_link_local_ipv6_mcast(const struct in6_addr *addr)
+{
+	return (((addr->s6_addr32[0] ^ htonl(0xff020000)) |
+		addr->s6_addr32[1] | addr->s6_addr16[6] | addr->s6_addr[14]) == 0);
+}
+
 static int br_multicast_ipv6_rcv(struct net_bridge *br,
 				 struct net_bridge_port *port,
 				 struct sk_buff *skb,
@@ -1710,9 +1727,10 @@ static int br_multicast_ipv6_rcv(struct net_bridge *br,
 		return 0;
 
 	/* Prevent flooding this packet if there is no listener present */
-	if (!ipv6_is_transient_multicast(&ip6h->daddr))
+	if (!ipv6_addr_is_ll_all_nodes(&ip6h->daddr) &&
+			!(is_solicited_node_mcast(&ip6h->daddr)) &&
+			!(is_link_local_ipv6_mcast(&ip6h->daddr)))
 		BR_INPUT_SKB_CB(skb)->mrouters_only = 1;
-
 	if (ip6h->nexthdr != IPPROTO_HOPOPTS ||
 	    ip6h->payload_len == 0)
 		return 0;
